cellxgene_schema_cli/cellxgene_schema/validate.py

```diff
@@ -0,0 +1,1588 @@
+                elif isinstance(x, ndarray):
+                    sparsity = 1 - count_nonzero(x) / float(cumprod(x.shape)[-1])
+                else:
+                    self.warnings.append(
+                        f"{x_name} matrix is of type {type(x)}, sparsity calculation hasn't been "
+                        f"implemented"
+                    )
+                    continue
+
+                if sparsity > max_sparsity:
+                    self.warnings.append(
+                        f"Sparsity of '{x_name}' is {sparsity} which is greater than {max_sparsity}, "
+                        f"and it is not a 'scipy.sparse.csr_matrix'. It is STRONGLY RECOMMENDED "
+                        f"to use this type of matrix for the given sparsity."
+                    )
+
+    def _validate_embedding_dict(self):
+
+        """
+        Validates the embedding dictionary -- it checks that all values of adata.obms are numpy arrays with the correct
+        dimension. Adds errors to self.errors if any. Checks that the keys start with "X_"
+
+        :rtype none
+        """
+
+        if not self.adata.obsm:
+            self.errors.append("No embeddings found in 'adata.obsm'")
+            return
+
+        obsm_with_x_prefix = 0
+        for key, value in self.adata.obsm.items():
+
+            if not isinstance(value, ndarray):
+                self.errors.append(
+                    f"All embeddings have to be of 'numpy.ndarray' type, "
+                    f"'adata.obsm['{key}']' is {type(value)}')"
+                )
+                continue
+
+            # Embeddings to be shown in cellxgene explorer
+            if key.startswith("X_"):
+                obsm_with_x_prefix += 1
+
+                if len(value.shape) < 2:
+                    self.errors.append(
+                        f"All embeddings must have as many rows as cells, and at least two columns."
+                        f"'adata.obsm['{key}']' has shape of '{value.shape}'"
+                    )
+                else:
+                    if value.shape[0] != self.adata.n_obs or value.shape[1] < 2:
+                        self.errors.append(
+                            f"All embeddings must have as many rows as cells, and at least two columns."
+                            f"'adata.obsm['{key}']' has shape of '{value.shape}'"
+                        )
+
+        if obsm_with_x_prefix == 0:
+            self.errors.append(
+                "At least one embedding in 'obsm' has to have a key with an 'X_' prefix."
+            )
+
+    def _are_children_of(
+        self, component: str, column: str, ontology_name: str, ancestors: List[str]
+    ) -> bool:
+
+        """
+        Checks if elements in the specified column of the component (e.g. 'assay_ontology_term_id' of 'adata.obs') are
+        children of the given ancestors.
+
+        Ancestors checks are inclusive, meaning that a value it's its own ancestor as well.
```

Comment:
```suggestion
        Ancestors checks are inclusive, meaning that a value is its own ancestor as well.
```

Reply:
fixed in 379bc27669a53de6f70eec8062a5adb9876a8203